# Practica 2 - Generación de Variables Aleatorias Continuas

In [1]:
import random
import math

# Ejercicio1 
Sea X una variable aleatoria exponencial con media 1. Proporcione un algoritmo eficiente para
simular una variable aleatoria cuya distribución es la distribución condicional de X dado que X <
0,05. Es decir, su función de densidad es 

Genere 1000 de estas variables y utilícelas para estimar E[X | X < 0,05]. Luego determine el
valor exacto de E[X | X < 0,05]

## Análisis:
Solo nos importan valores dentro del rango 0 < X < 0.05, que cumplen la condición. Debemos de encontrar la función acumulada:
 
 $$F(x) = \frac{1 - e^{-x}}{1 - e^{-0.05}}$$
para $0 < x < 0.05$

Ahora generamos una variable aleatoria uniforme $U$ en el intervalo $(0,1)$ y la transformamos usando la función inversa de $F(x)$:

$$U \sim \text{Uniforme}(0,1)$$

Para encontrar $X = F^{-1}(U)$, despejamos $x$ de la ecuación $U = F(x)$:

$$U = \frac{1 - e^{-x}}{1 - e^{-0.05}}$$

$$U(1 - e^{-0.05}) = 1 - e^{-x}$$

$$e^{-x} = 1 - U(1 - e^{-0.05})$$

$$-x = \ln[1 - U(1 - e^{-0.05})]$$

$$X = -\ln[1 - U(1 - e^{-0.05})]$$

## Explicación del algoritmo: 
1. Generar U uniforme en (0,1)
2. Calcular X usando la fórmula $X = -\ln[1 - U(1 - e^{-0.05})]$
3. Retornar X como la variable aleatoria simulada.

y ahora solo se generan 1000 de estas variables para estimar $E[X | X < 0.05]$ y luego determinar el valor exacto de $E[X | X < 0.05]$.

    

In [8]:
n = 1000
suma = 0

# el valor esperado de X dado que X < 0.05 se puede estimar 
# generando n variables aleatorias uniformes U en el
# intervalo (0,1) y transformándolas usando la función inversa de F(x):
for i in range(n):
    u = random.random()

    x = -math.log(1 - u * (1 - math.exp(-0.05)))

    suma += x

promedio = suma / n

print("Estimación de E[X | X < 0.05]:", promedio)

# valor exacto de E[X | X < 0.05] se puede calcular usando la integral de la función de densidad condicional:
exacto = (1 - 1.05 * math.exp(-0.05)) / (1 - math.exp(-0.05))
print("Valor exacto de E[X | X < 0.05]:", exacto)

Estimación de E[X | X < 0.05]: 0.024970168303712858
Valor exacto de E[X | X < 0.05]: 0.02479167534670473


# Ejercicio 2
Suponga que es relativamente fácil generar variables aleatorias a partir de cualquiera de las dis-
tribuciones Fi, i = 1, . . . , n. ¿Cómo podríamos generar una variable aleatoria con función de distribución

$$F(x) = \sum_{i=1}^{n} p_i(x)$$

donde pi, i = 1, . . . , n, son números no negativos cuya suma es 1?

## Análisis: 

Generamos una distribución uniforme U ~ Uniforme(0,1) y contruimos intervalos acumulados correpondiendo a cada una de las distribuciones Fi, i = 1, . . . , n. Luego, dependiendo del valor de U, seleccionamos la distribución correspondiente y generamos una variable aleatoria a partir de esa distribución.

Si tuvieramos U = 0.37 y con intervalos [0,0.2) para F1, [0.2,0.5) para F2 y [0.5,1) para F3, entonces U = 0.37 cae en el intervalo de F2, por lo que generamos una variable aleatoria a partir de la distribución F2.

un algoritmo para esto sería: 
1. Generar U uniforme en (0,1)
2. Determinar el índice I que:
   - p1 + p2 + ... + p(I-1) < U <= p1 + p2 + ... + pI
3. Una vez seleccionado I, genere X usando la distribución Fi. 
4. retornar x

Si las probabilidades son del 20%, 30% y 50% respectivamente, la línea de tiempo se divide de la siguiente manera:

```text
0       0.20     0.50                             1.00

|--------|--------|--------------------------------|
    F1       F2                   F3
   20%      30%                  50%
```

Si el número fuera 0.10 cae en F1, si fuera 0.37 cae en F2 y si fuera 0.75 cae en F3.

## Algoritmo

```text
Generar U ~ Uniforme(0, 1)
acumulada = 0

Para i = 1 hasta n:
    acumulada = acumulada + p_i
    
    si U <= acumulada entonces
        generar X usando Fi
        retornar X
```


# Ejercicio 3

Utilizando el resultado del Ejercicio 2, proporcione algoritmos para generar variables aleatorias a partir de las siguientes distribuciones.

(a) $$F(x) = \frac{x + x^3 + x^5}{3}, \quad 0 \le x \le 1$$

(b) $$F(x) = \begin{cases} \frac{1 - e^{-2x} + 2x}{3} & 0 < x < 1 \\ \frac{3 - e^{-2x}}{3} & 1 \le x < \infty \end{cases}$$

(c) $$F(x) = \sum_{i=1}^{n} \alpha_i x^i, \quad 0 \le x \le 1, \quad \text{donde } \alpha_i \ge 0, \quad \sum_{i=1}^{n} \alpha_i = 1$$


(a) en este podemos dividir los términos en tres partes. 
 $$F(x)= \frac{1}{3}x + \frac{1}{3}x^3 + \frac{1}{3}x^5$$
 ahora ya tiene la forma de la distribución que se puede usar para generar variables aleatorias. utilizando:
 p1 = p2 = p3 = 1/3 y también F1(x) = x, F2(x) = x^3, F3(x) = x^5.

 Ya con esto podemos escoger una de las tres distribuciones y generar una variable aleatoria a partir de ella.
 U1 ~ U(0,1) y dividirlo entre 0, 1/3,2/3,1. 

 Entonces:
- si U1 <= 1/3, usamos F1;
- si 1/3 < U1 <= 2/3, usamos F2;
- si U1 > 2/3, usamos F3.
Pero todavía falta generar el valor X.

## algoritmo
```text
Generar U1 ~ Uniforme(0, 1)
Generar U2 ~ Uniforme(0, 1)
Si U1 <= 1/3 entonces
    X = U2
Si 1/3 < U1 <= 2/3 entonces
    X = U2^(1/3)
Si U1 > 2/3 entonces
    X = U2^(1/5)

retornar X 
```

### (b)

En este caso podemos identificar que la función está formada por dos distribuciones diferentes:

$$
F(x)=
\begin{cases}
\frac{1-e^{-2x}+2x}{3}, & 0<x<1 \\
\frac{3-e^{-2x}}{3}, & x>1
\end{cases}
$$

Para $0<x<1$ podemos separarla así:

$$
F(x)=\frac{1}{3}(1-e^{-2x})+\frac{2}{3}x
$$

Entonces podemos identificar:

$$
F_1(x)=1-e^{-2x}
$$

que corresponde a una distribución exponencial con tasa:

$$
\lambda=2
$$

y también:

$$
F_2(x)=x
$$

que corresponde a una distribución uniforme entre 0 y 1.

Por lo tanto usamos:

$$
p_1=\frac{1}{3}
$$

$$
p_2=\frac{2}{3}
$$

Primero generamos $U_1 \sim U(0,1)$ para escoger cuál distribución utilizar.

Entonces:

- si $U_1 \leq \frac{1}{3}$, usamos $F_1$;
- si $U_1 > \frac{1}{3}$, usamos $F_2$.

Para generar la exponencial de tasa 2 usamos transformación inversa:

$$
X=-\frac{1}{2}\ln(U_2)
$$

y para la uniforme simplemente:

$$
X=U_2
$$

## algoritmo

```text
Generar U1 ~ Uniforme(0, 1)
Generar U2 ~ Uniforme(0, 1)

Si U1 <= 1/3 entonces
    X = -(1/2) ln(U2)

Si U1 > 1/3 entonces
    X = U2

retornar X
```


### (c)

En este caso tenemos una forma más general:

$$
F(x)=\sum_{i=1}^{n}\alpha_i x^i
$$

donde:

$$
\alpha_i \geq 0
$$

y:

$$
\sum_{i=1}^{n}\alpha_i=1
$$

Podemos considerar cada término como una distribución diferente:

$$
F_i(x)=x^i
$$

y cada distribución se escoge con probabilidad:

$$
p_i=\alpha_i
$$

Primero escogemos un índice $i$ utilizando las probabilidades:

$$
\alpha_1,\alpha_2,\ldots,\alpha_n
$$

Después generamos:

$$
U \sim U(0,1)
$$

Como:

$$
F_i(x)=x^i
$$

usamos transformación inversa:

$$
U=X^i
$$

y despejando $X$:

$$
X=U^{1/i}
$$

De esta forma podemos generar una variable aleatoria utilizando el término seleccionado.

## algoritmo

```text
Escoger un índice i usando las probabilidades
α1, α2, ..., αn

Generar U ~ Uniforme(0, 1)

X = U^(1/i)

retornar X

# Ejercicio 4 
Una compañía de seguros contra siniestros tiene 1000 asegurados, cada uno de los cuales presentará una reclamación en el próximo mes de manera independiente con probabilidad 0.05. 

Suponiendo que los montos de las reclamaciones son variables aleatorias exponenciales independientes con media $800, utilice simulación para estimar la probabilidad de que la suma de estas reclamaciones exceda $50,000

## análisis: 

En este ejercicio tenemos una compañía con **1000 asegurados**.

Cada asegurado puede presentar una reclamación con probabilidad:

$$
p = 0.05
$$

Esto significa que para cada asegurado podemos generar un número aleatorio:

$$
U \sim U(0,1)
$$

y si:

$$
U < 0.05
$$

entonces consideramos que esa persona sí presenta una reclamación.

Si presenta una reclamación, el monto sigue una distribución exponencial con media:

$$
800
$$

Para una distribución exponencial:

$$
E[X] = \frac{1}{\lambda}
$$

por lo tanto:

$$
\lambda = \frac{1}{800}
$$

Para generar el monto de una reclamación podemos usar transformación inversa:

$$
X = -800\ln(U)
$$

donde $U$ es otro número uniforme entre 0 y 1.

La idea de la simulación es recorrer los 1000 asegurados, verificar cuáles presentan reclamación y sumar todos los montos generados.

Al terminar tenemos el total de reclamaciones de un posible mes.

Después verificamos si:

$$
\text{total} > 50000
$$

Pero como queremos estimar una probabilidad, no basta con hacer una sola simulación.  
Debemos repetir todo el proceso muchas veces, por ejemplo 10000 veces.

En cada simulación contamos si el total superó los \$50,000.

Finalmente estimamos la probabilidad con:

$$
P(\text{total} > 50000)
\approx
\frac{\text{cantidad de simulaciones donde el total superó 50000}}
{\text{cantidad total de simulaciones}}
$$

## Algoritmo

```text
repetir muchas simulaciones

    total = 0

    para cada uno de los 1000 asegurados

        generar U ~ Uniforme(0,1)

        si U < 0.05 entonces

            generar otro U ~ Uniforme(0,1)

            monto = -800 ln(U)

            sumar monto al total

    si total > 50000 entonces
        aumentar contador

probabilidad = contador / número de simulaciones
```

En promedio se esperan:

$$
1000(0.05)=50
$$

reclamaciones por mes.

Como cada reclamación tiene una media de \$800, el total promedio esperado sería aproximadamente:

$$
50(800)=40000
$$

Por eso tiene sentido que el ejercicio pregunte por la probabilidad de superar los \$50,000, ya que es un valor mayor al total promedio esperado.

In [11]:
# usar simulación para P(suma de reclamos > 50000)
simulaciones = 10000
exceden_50000 = 0

for i in range(simulaciones):
    total = 0 
    for aseguradora in range(1000):
        # random reclamo o no 
        u = random.random()
        if u < 0.05:
            # generar reclamo exponencial con media 800
            u_monto = random.random()
            monto = -800 * math.log(u_monto)
            total += monto
    # revisa si la suma excede 50000
    if total > 50000:
        exceden_50000 += 1

probabilidad = exceden_50000 / simulaciones
print("Estimación de P(suma de reclamos > 50000):", probabilidad)

Estimación de P(suma de reclamos > 50000): 0.1055


Los montos se suman y se verifica si el total supera los \$50,000. Se repite el proceso muchas veces para estimar la probabilidad de que esto ocurra. Tuvimos como resultado que la probabilidad de que la suma de las reclamaciones exceda \$50,000 es aproximadamente 0.1055, lo que significa que hay un 10.55% de probabilidad de que esto ocurra en un mes dado.

# Ejercicio 5 
Escriba un programa que genere variables aleatorias normales utilizando el método del Ejemplo
5f (método de rechazo con una distribución exponencial de tasa 1).

En este ejercicio queremos generar una variable aleatoria normal estándar:

$$
Z \sim N(0,1)
$$

El método usa **aceptación y rechazo** tomando como distribución auxiliar una exponencial de tasa 1:

$$
Y \sim Exp(1)
$$

Primero generamos un valor exponencial:

$$
Y = -\ln(U_1)
$$

donde:

$$
U_1 \sim U(0,1)
$$

Luego generamos otro número uniforme:

$$
U_2 \sim U(0,1)
$$

El valor de $Y$ se acepta solamente si cumple:

$$
U_2 \leq e^{-\frac{(Y-1)^2}{2}}
$$

Si no cumple esta condición, se rechaza el valor y se vuelve a generar otro.

Una vez que tenemos un valor aceptado, todavía necesitamos decidir si la normal será positiva o negativa.

Para esto generamos otro número:

$$
U_3 \sim U(0,1)
$$

Entonces:

- si $U_3 \leq 0.5$, usamos:

$$
Z=Y
$$

- si $U_3 > 0.5$, usamos:

$$
Z=-Y
$$

De esta forma obtenemos una variable:

$$
Z \sim N(0,1)
$$

## Algoritmo

```text
repetir

    generar U1 ~ Uniforme(0,1)

    Y = -ln(U1)

    generar U2 ~ Uniforme(0,1)

hasta que:

    U2 <= exp(-(Y - 1)^2 / 2)


generar U3 ~ Uniforme(0,1)

si U3 <= 0.5 entonces
    Z = Y

si U3 > 0.5 entonces
    Z = -Y

retornar Z
```

La idea del método es generar primero valores positivos usando una distribución exponencial y luego aceptar solamente aquellos que cumplen la condición de rechazo.

Después se escoge aleatoriamente el signo positivo o negativo para obtener finalmente una variable normal estándar.

Si queremos comprobar que el método funciona, podemos generar muchas variables y verificar que su media sea aproximadamente:

$$
E[Z] \approx 0
$$

y su varianza aproximadamente:

$$
Var(Z) \approx 1
$$

In [12]:
# definimos la función para generar variables aleatorias 
# normales usando el método de rechazo con una distribución 
# exponencial de tasa 1
def normal_rechazo():

    while True:
        # generamos random u1 y calculamos y = -ln(u1)
        u1 = random.random()
        y = -math.log(u1)

        u2 = random.random()
        
        if u2 <= math.exp(-((y - 1) ** 2) / 2):
            break

    u3 = random.random()
    # verificamos si u3 <= 0.5 para determinar el signo de z
    if u3 <= 0.5:
        z = y
    else:
        z = -y

    return z


print(normal_rechazo())

0.8997024263487584


Con este resultado de normalidad podemos notar un resultado de 0.899 lo que significa que se generó una observacion con ese normal estándar. 
Esto nos demuestra que el método de rechazo con una distribución exponencial de tasa 1 es efectivo para generar variables aleatorias normales estándar.

# Ejercicio 6
Escriba un programa que genere las primeras T unidades de tiempo de un proceso de Poisson
con tasa λ

### Proceso de Poisson 
Eventos que ocurren aleatoriamente en un tiempo determinado.
Para este caso tenemos que la tada es de 2 eventos por unidad de tiempo. 
 X ~ Exp(λ)

 ```text
 t = 0 
 mientras t < T: 
    generar U ~ Uniforme(0,1)
    generar X = -ln(U)/λ
    t = t + X
    
    si t <=T:
        guardar t

 ```



In [16]:
lam = 2 
T = 20 

t = 0 
eventos = []

while t < T:
    U = random.uniform(0, 1)
    X = -(1/lam) * math.log(U)
    t = t + X

    if t <= T:
        eventos.append(t)

print(eventos)

[0.11399705929534305, 0.42339855707553165, 1.0024100270851783, 2.4132876328962016, 2.5219186638605353, 3.007912537019478, 3.724005886058076, 3.811689710736937, 4.267055259496884, 4.573840017279412, 4.7221492955733115, 4.854446921468656, 5.574466745235318, 6.649230028170317, 6.758305660023075, 7.07832540496597, 7.713825599688212, 7.92236015235491, 8.204579556302914, 8.329741527192931, 9.46053645361757, 9.62923797858386, 10.017053375690027, 10.809370933072927, 11.064570331094775, 11.364207146919325, 11.732577425511213, 13.588242666279509, 13.592863153050855, 13.643172876394612, 13.851869158828608, 14.021512512799855, 14.831911757684507, 14.993722519677505, 15.636641735126073, 15.754091615363373, 15.808651584550296, 16.032986374382457, 17.362794704317274, 17.799303673468803, 17.983453357126898, 18.043345281119848, 19.062730043264466, 19.20086153728129, 19.992847047180486]


estos resultados no son los tiempos entre eventos, sino los momentos en los que ocurrió cada evento. Por ejemplo, si el primer evento ocurre en t=0.5 y el segundo en t=1.2, entonces los tiempos entre eventos serían 0.5 y 0.7, respectivamente.

Esto lo que significa es que al generar la exponencial con tasa 2, estamos generando los tiempos entre eventos, y al ir sumando esos tiempos obtenemos los momentos en los que ocurren los eventos.